login-utils/lslogins.c

```diff
@@ -524,10 +532,63 @@ static int parse_utmpx(const char *path, size_t *nrecords, struct utmpx **record
 	return -EINVAL;
 }
 
-static void get_lastlog(struct lslogins_control *ctl, uid_t uid, void *dst, int what)
+#ifdef HAVE_LASTLOG2
+static int get_lastlog2(struct lslogins_control *ctl, const char *user, void *dst, int what)
+{
+	switch (what) {
+	case LASTLOG_TIME: {
+		time_t *t = dst;
+		int64_t res_time = 0;
+
+		if (ll2_read_entry(ctl->lastlog2_path, user, &res_time, NULL, NULL, NULL, NULL) != 0)
```

Comment:
Does not have the library better API? 

I mean something not based on the path. It seems it opens the lastlog2 file always when an entry is required. It would be better to have something g file-descriptors based.

Reply:
The library API is optimized for pam_lastlog2 and lastlog2, there is no optimized API for lslogins. Currently it looks like lslogins is the only 3rd party application using directly liblastlog2.
For the other applications a callback function is used, which avoids the many reads, but in this case, this would mean lslogins would need to read all entries into memory.